### PR TITLE
Envoy rate limit descriptor fix

### LINKED ISSUE #10124: Envoy skips calling rate limiting service when request header is not present
We have configured the rate limiting service with three request headers actions like 

```
        "rate_limits": [
         {
          "stage": 0,
          "actions": [
           {
            "source_cluster": {}
           },
           {
            "destination_cluster": {}
           },
           {
            "request_headers": {
             "header_name": ":authority",
             "descriptor_key": "authority"
            }
           },
           {
            "request_headers": {
             "header_name": ":path",
             "descriptor_key": "path"
            }
           },
           {
            "request_headers": {
             "header_name": "custom-header",
             "descriptor_key": "custom"
            }
           },
          ]
         }
        ]
       }
      ]
```
If in a request the `custom-header` does not have a value, Envoy skips calling the rate limiting service.

Debugging it further, found that if any of the `populateDescriptor` call here returns false it skips adding the descriptor

https://github.com/envoyproxy/envoy/blob/5be15b31aa46e3c3776a15fd3d399c67ac432598/source/common/router/router_ratelimit.cc#L121

and 

 `populateDescriptor`  implementation of `RequestHeadersAction` returns false if there is no value

https://github.com/envoyproxy/envoy/blob/5be15b31aa46e3c3776a15fd3d399c67ac432598/source/common/router/router_ratelimit.cc#L41

Since the descriptor list is empty, rate limit service is not called.

Is this expected behaviour? Should we not call rate limiting service with two headers as descriptor entries which have values rather than completely skipping it? 

Should we simply add a descriptor if it has atleast one entry?

### PR DESCRIPTION
**Pull request by rohan07**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: It resolves - https://github.com/envoyproxy/envoy/issues/10124 indirectly by adding an extra config flag to RequestHeaders through which it is possible for descriptors to be sent on a partial match
Risk Level: low - optional feature
Testing: unit tests added
Docs Changes: Added
Release Notes: Added
[Optional Fixes #Issue] Fixes #10124 
[Optional Deprecated:]

### CODE REVIEW

**Comment by rohan07** [comment]
Need to add associated documentation changes for the new configuration option. Opening PR so that @ramaraochavali can view it.

File: api/envoy/config/route/v3/route_components.proto [hunk 1]
```diff
@@ -1385,6 +1385,10 @@ message RateLimit {
 
       // The key to use in the descriptor entry.
       string descriptor_key = 2 [(validate.rules).string = {min_bytes: 1}];
+
+      // The optional parameter which determines if the RLS is called if a request
```
Comment (ramaraochavali) [inline api/envoy/config/route/v3/route_components.proto:1389]: If set to true, Envoy skips the descriptor while calling rate limiting service when header is not present in the request. Defaults to false, which skips calling rate limiting service if this header is not present in the request.

File: source/common/router/router_ratelimit.cc [hunk 2]
```diff
@@ -38,10 +38,14 @@ bool RequestHeadersAction::populateDescriptor(const Router::RouteEntry&,
                                               const Http::HeaderMap& headers,
                                               const Network::Address::Instance&) const {
   const Http::HeaderEntry* header_value = headers.get(header_name_);
-  if (!header_value) {
+  if (!header_value && !skip_if_absent_) {
```
Comment (ramaraochavali) [inline source/common/router/router_ratelimit.cc:41]: nit: Simplify it like this and add a comment
```
  if (!header_value) {
    return !skip_if_absent_;
  }
```

File: test/common/router/router_ratelimit_test.cc [hunk 3]
```diff
@@ -366,6 +366,94 @@ TEST_F(RateLimitPolicyEntryTest, RequestHeaders) {
               testing::ContainerEq(descriptors_));
 }
 
+// Validate that a descriptor is added if at least one request header has value.
+TEST_F(RateLimitPolicyEntryTest, RequestHeadersPartialMatchWithSkipIfAbsent) {
```
Comment (ramaraochavali) [inline test/common/router/router_ratelimit_test.cc:370]: /s/RequestHeadersPartialMatchWithSkipIfAbsent/RequestHeadersWithSkipIfAbsent

File: test/common/router/router_ratelimit_test.cc [hunk 4]
```diff
@@ -366,6 +366,94 @@ TEST_F(RateLimitPolicyEntryTest, RequestHeaders) {
...
+  setupTest(yaml);
+  Http::TestHeaderMapImpl header{{"x-header-name", "test_value"}};
+
+  rate_limit_entry_->populateDescriptors(route_, descriptors_, "service_cluster", header,
+                                         default_remote_address_);
+  EXPECT_THAT(std::vector<Envoy::RateLimit::Descriptor>({{{{"my_header_name", "test_value"}}}}),
+              testing::ContainerEq(descriptors_));
+}
+
+TEST_F(RateLimitPolicyEntryTest, RequestHeadersPartialMatchWithoutSkipIfAbsent) {
```
Comment (ramaraochavali) [inline test/common/router/router_ratelimit_test.cc:392]: /s/RequestHeadersPartialMatchWithoutSkipIfAbsent/RequestHeadersWithDefaultSkipIfAbsent and add a comment above the test

File: test/common/router/router_ratelimit_test.cc [hunk 5]
```diff
@@ -366,6 +366,94 @@ TEST_F(RateLimitPolicyEntryTest, RequestHeaders) {
...
+  Http::TestHeaderMapImpl header{{"x-header-name", "test_value"}, {"x-header", "test_value"}};
+
+  rate_limit_entry_->populateDescriptors(route_, descriptors_, "service_cluster", header,
+                                         default_remote_address_);
+  EXPECT_THAT(std::vector<Envoy::RateLimit::Descriptor>(
+                  {{{{"my_header_name", "test_value"}, {"my_header", "test_value"}}}}),
+              testing::ContainerEq(descriptors_));
+}
+
+TEST_F(RateLimitPolicyEntryTest, RequestHeadersMatchDefaultSkipIfAbsent) {
```
Comment (ramaraochavali) [inline test/common/router/router_ratelimit_test.cc:436]: This is not needed

File: test/common/router/router_ratelimit_test.cc [hunk 6]
```diff
@@ -366,6 +366,94 @@ TEST_F(RateLimitPolicyEntryTest, RequestHeaders) {
...
+
+  setupTest(yaml);
+  Http::TestHeaderMapImpl header{{"x-header-test", "test_value"}};
+
+  rate_limit_entry_->populateDescriptors(route_, descriptors_, "service_cluster", header,
+                                         default_remote_address_);
+  EXPECT_TRUE(descriptors_.empty());
+}
+
+TEST_F(RateLimitPolicyEntryTest, RequestHeadersCompleteMatch) {
```
Comment (ramaraochavali) [inline test/common/router/router_ratelimit_test.cc:413]: This also should have been tested already

**Comment by ramaraochavali** [comment]
Add a version history entry https://github.com/envoyproxy/envoy/blob/master/docs/root/version_history/current.rst in alpha order

File: docs/root/version_history/current.rst [hunk 7]
```diff
@@ -39,10 +39,11 @@ Changes
 * network filters: added a :ref:`rocketmq proxy filter <config_network_filters_rocketmq_proxy>`.
 * prometheus stats: fix the sort order of output lines to comply with the standard.
 * request_id: added to :ref:`always_set_request_id_in_response setting <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.always_set_request_id_in_response>`
-  to set :ref:`x-request-id <config_http_conn_man_headers_x-request-id>` header in response even if
+  to set :ref:`x-request-id <config_http_conn_man_headers_x-request-id>` header in response even if skip_if_absent
```
Comment (ramaraochavali) [inline docs/root/version_history/current.rst:42]: not needed change looks like?

Comment (rohan07) [reply docs/root/version_history/current.rst:42]: Yup, pushing the fix, accidental change

File: api/envoy/config/route/v3/route_components.proto [hunk 8]
```diff
@@ -1385,6 +1385,11 @@ message RateLimit {
 
       // The key to use in the descriptor entry.
       string descriptor_key = 2 [(validate.rules).string = {min_bytes: 1}];
+
+      // If set to true, Envoy skips the descriptor while calling rate limiting service
+      // when header is not present in the request. Defaults to false, which skips calling
```
Comment (htuch) [inline api/envoy/config/route/v3/route_components.proto:1390]:
```suggestion
      // when header is not present in the request. By default this skips calling
```
i.e. it's redundant to say that the default of a primitive proto3 type is false, but your description of the actual behavior is very helpful.

File: docs/root/version_history/current.rst [hunk 9]
```diff
@@ -43,6 +43,7 @@ Changes
   tracing is not forced.
 * router: allow retries of streaming or incomplete requests. This removes stat `rq_retry_skipped_request_not_complete`.
 * router: allow retries by default when upstream responds with :ref:`x-envoy-overloaded <config_http_filters_router_x-envoy-overloaded_set>`.
+* router: allow Rate Limiting service to be called in case of missing request header for a descriptor if the :ref:`skip_if_absent <envoy_v3_api_field_config.route.v3.RateLimit.Action.RequestHeaders.skip_if_absent>` field is set to true.
```
Comment (htuch) [inline docs/root/version_history/current.rst:46]: Nit: Rate Limiting SErvice

File: api/envoy/config/route/v3/route_components.proto (hunk 8 above)
Comment (rohan07) [reply api/envoy/config/route/v3/route_components.proto]: Sounds good, @htuch! Pushed the changes as you suggested! Thanks!

**Comment by ramaraochavali** [comment]
LGTM. @mattklein123  @junr03 Can you PTAL?

File: source/common/router/router_ratelimit.cc [hunk 10]
```diff
@@ -38,10 +38,11 @@ bool RequestHeadersAction::populateDescriptor(const Router::RouteEntry&,
                                               const Http::HeaderMap& headers,
                                               const Network::Address::Instance&) const {
   const Http::HeaderEntry* header_value = headers.get(header_name_);
+
+  // If header is not present and skip_if_absent_ is not set to true, short circuit here and return.
```
Comment (mattklein123) [inline source/common/router/router_ratelimit.cc:42]: nit: this comment is not quite accurate, as you return no matter what. Can you clarify?

Comment (rohan07) [reply source/common/router/router_ratelimit.cc:42]: Hello @mattklein123 - Yes, you are right, I agree it can be misleading. Let me rephrase it. What I was going for was "do not add the descriptor".  Thanks for pointing it out!

**Comment by mattklein123** [comment]
@rohan07 please *never* force push. It makes reviews much more difficult. Thanks!

File: source/common/router/router_ratelimit.cc [hunk 11]
```diff
@@ -38,10 +38,11 @@ bool RequestHeadersAction::populateDescriptor(const Router::RouteEntry&,
                                               const Http::HeaderMap& headers,
                                               const Network::Address::Instance&) const {
   const Http::HeaderEntry* header_value = headers.get(header_name_);
+
+  // If header is not present and skip_if_absent_ is true, ignore the descriptor
```
Comment (mattklein123) [inline source/common/router/router_ratelimit.cc:42]: This comment is still not correct. Returning has no bearing on whether skip_if_absent_ is true?

**Comment by mattklein123** [comment]
Thanks please merge master and force push to fix DCO.

/wait

**Comment by rohan07** [comment]
Apologies @mattklein123 - My previous force push was also because of the DCO. I was experiencing some performance issues with my old machine so had to get a new one.

**Comment by rohan07** [comment]
Hi @mattklein123 - Apologies again, it wasn't allowing me to sign off on my previous commit using my new machine for some reason, it worked from my previous machine. Thanks for your help!

**Comment by rohan07** [comment]
Hi @mattklein123 - I am seeing some of the other PR's which are open facing the same issue with circleci/coverage failing (with 96.7%, lower than the 97% threshold). I believe during my earlier commits, the particular test was passing. Can you please advise the future steps I should take?
